fvm/evm/offchain/sync/replayer.go

```diff
@@ -0,0 +1,65 @@
+package sync
+
+import (
+	gethTracers "github.com/onflow/go-ethereum/eth/tracers"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/fvm/evm/events"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// Replayer consumes EVM transaction and block
+// events, re-execute EVM transaction and follows EVM chain.
+// this allows using different tracers and storage solutions.
+type Replayer struct {
+	chainID         flow.ChainID
+	rootAddr        flow.Address
+	logger          zerolog.Logger
+	storageProvider types.StorageProvider
+	tracer          *gethTracers.Tracer
+	validateResults bool
+}
+
+// NewReplayer constructs a new Replayer
+func NewReplayer(
+	chainID flow.ChainID,
+	rootAddr flow.Address,
+	sp types.StorageProvider,
+	logger zerolog.Logger,
+	tracer *gethTracers.Tracer,
+	validateResults bool,
+) *Replayer {
+	return &Replayer{
+		chainID:         chainID,
+		rootAddr:        rootAddr,
+		storageProvider: sp,
+		logger:          logger,
+		tracer:          tracer,
+		validateResults: validateResults,
+	}
+}
+
+// OnBlockReceived is called when a new block is received
+// (including all the related transaction executed events)
+func (cr *Replayer) OnBlockReceived(
+	transactionEvents []events.TransactionEventPayload,
+	blockEvent *events.BlockEventPayload,
+) (types.ReplayResults, error) {
+	// prepare storage
+	st, err := cr.storageProvider.GetSnapshotAt(blockEvent.Height)
```

Comment:
Does this assumes the block are received in sequential order?

If so, please comment with the assumptions.

If the storage doesn't have the snapshot at the given height, should we return error? 

Is the next `OnBlockReceived` call got blocked until this call is finished?

Reply:
Its not expecting to be called sequential, that actually part of reason I made the storageProvider stand out and return the deltas on a call. For example, we could eventually concurrently call OnBlock executed using the registers we have on chunk data pack.
Having said that, I noticed I need to make the blocks its own thing as well, so we don't need any sequential calls.

Reply:
I added some comments, I noticed this currently has to be sequentially called but has the option to be concurrent in the future, when we move the block context provider logic to an external code.